### PR TITLE
eksctl install flux now handles re-installation

### LINKED ISSUE #1084: Handle "flux already installed in cluster" situation
If flux is already installed in the cluster eksctl should handle it earlier.

```
$ EKSCTL_EXPERIMENTAL=true eksctl install flux -f examples/01-simple-cluster.yaml --git-url=git@github.com:martina-if/flux-test-2.git --git-email=martina+flux@weave.works
[ℹ]  Cloning git@github.com:martina-if/flux-test-2.git
Cloning into '/tmp/eksctl-install-flux-clone194001168'...
warning: templates not found /home/martina/.git_template
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (3/3), done.
[ℹ]  Writing Flux manifests
[ℹ]  Installing Flux into the cluster
[ℹ]  replaced "flux:Deployment.apps/flux"
[ℹ]  replaced "flux:Secret/flux-git-deploy"
[ℹ]  replaced "flux:Deployment.apps/memcached"
[✖]  You may find the local clone of git@github.com:martina-if/flux-test-2.git used by eksctl at /tmp/eksctl-install-flux-clone194001168
[✖]  Service "memcached" is invalid: spec.clusterIP: Invalid value: "": field is immutable
```

```
$ eksctl version
[ℹ]  version.Info{BuiltAt:"1564746220", GitCommit:"0.2.0-46-g35bedce0", GitTag:""}
```

### PR DESCRIPTION
**Pull request by marccarre**
### Description

`eksctl install flux` now handles re-installation by deleting all resources and creating them again.

Other approaches were considered, like simply using `kubectl apply`
but given this would add an external dependency, this solution was
chosen for now.

Also note that we prompt the user before going ahead and deleting all-the-things. See also "_Manual test_" below.

Fixes #1084.

### Checklist

- [x] Code compiles correctly (i.e `make build`)
- [x] ~Added tests that cover your change (if possible)~ I didn't add any E2E test for this given the cost associated to them, and given this is an edge case. Also note that there could be a need to mock `stdin` to simulate the user's input, which is feasible with some additional boilerplate, if we nevertheless want to proceed with an E2E test in the future.
- [x] All unit tests passing (i.e. `make test`)
- [x] ~Added/modified documentation as required (such as the `README.md`, and `examples` directory)~
- [x] Manually tested

### Manual test

```console
$ EKSCTL_EXPERIMENTAL=true AWS_PROFILE=default-mfa ./eksctl \
    -f examples/01-simple-cluster.yaml \
    install flux \
    --git-url=git@github.com:marccarre/my-gitops-repo.git \
    --git-email=carre.marc@gmail.com

[ℹ]  Namespace "flux" already exists. eksctl will now delete all its content before proceeding to install Flux.
Continue? [y/n]:
Invalid choice.
Continue? [y/n]: meh
Invalid choice.
Continue? [y/n]: n
[ℹ]  Installation of Flux was aborted.
```

```console
$ EKSCTL_EXPERIMENTAL=true AWS_PROFILE=default-mfa ./eksctl \
    -f examples/01-simple-cluster.yaml \
    install flux \
    --git-url=git@github.com:marccarre/my-gitops-repo.git \
    --git-email=carre.marc@gmail.com

[ℹ]  Namespace "flux" already exists. eksctl will now delete all its content before proceeding to install Flux.
Continue? [y/n]: y
[ℹ]  deleted namespace flux
[ℹ]  waiting for pods under namespace flux to be deleted
[ℹ]  waiting for pods under namespace flux to be deleted
[ℹ]  waiting for pods under namespace flux to be deleted
[ℹ]  waiting for pods under namespace flux to be deleted
[ℹ]  waiting for namespace flux to be deleted
[ℹ]  waiting for namespace flux to be deleted
[ℹ]  waiting for namespace flux to be deleted
[ℹ]  waiting for namespace flux to be deleted
[ℹ]  Generating public key infrastructure for the Helm Operator and Tiller
[ℹ]    this may take up to a minute, please be patient
[!]  Public key infrastructure files were written into directory "/var/folders/24/d3mml6bn20nftpt91cfldq1h0000gn/T/eksctl-helm-pki937905532"
[!]  please move the files into a safe place or delete them
[ℹ]  Generating manifests
[ℹ]  Cloning git@github.com:marccarre/my-gitops-repo.git
Cloning into '/var/folders/24/d3mml6bn20nftpt91cfldq1h0000gn/T/eksctl-install-flux-clone-816069291'...
remote: Enumerating objects: 142, done.
remote: Total 142 (delta 0), reused 0 (delta 0), pack-reused 142
Receiving objects: 100% (142/142), 55.70 KiB | 160.00 KiB/s, done.
Resolving deltas: 100% (40/40), done.
[ℹ]  Writing Flux manifests
[ℹ]  Applying manifests
[ℹ]  created "Namespace/flux"
[ℹ]  created "flux:Deployment.apps/memcached"
[ℹ]  created "flux:Service/memcached"
[ℹ]  created "flux:Deployment.extensions/tiller-deploy"
[ℹ]  replaced "CustomResourceDefinition.apiextensions.k8s.io/helmreleases.helm.fluxcd.io"
[ℹ]  created "flux:Deployment.apps/flux"
[ℹ]  created "flux:Deployment.apps/flux-helm-operator"
[ℹ]  created "flux:ConfigMap/flux-helm-tls-ca-config"
[ℹ]  created "flux:Secret/flux-git-deploy"
[ℹ]  created "flux:Service/tiller-deploy"
[ℹ]  created "flux:ServiceAccount/tiller"
[ℹ]  replaced "ClusterRoleBinding.rbac.authorization.k8s.io/tiller"
[ℹ]  created "flux:ServiceAccount/helm"
[ℹ]  created "flux:Role.rbac.authorization.k8s.io/tiller-user"
[ℹ]  replaced "kube-system:RoleBinding.rbac.authorization.k8s.io/tiller-user-binding"
[ℹ]  created "flux:ServiceAccount/flux"
[ℹ]  replaced "ClusterRole.rbac.authorization.k8s.io/flux"
[ℹ]  replaced "ClusterRoleBinding.rbac.authorization.k8s.io/flux"
[ℹ]  created "flux:ServiceAccount/flux-helm-operator"
[ℹ]  replaced "ClusterRole.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  replaced "ClusterRoleBinding.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  Applying Helm TLS Secret(s)
[ℹ]  created "flux:Secret/flux-helm-tls-cert"
[ℹ]  created "flux:Secret/tiller-secret"
[!]  Note: certificate secrets aren't added to the Git repository for security reasons
[ℹ]  Waiting for Helm Operator to start
ERROR: logging before flag.Parse: E0827 19:36:51.844610   16191 portforward.go:331] an error occurred forwarding 58983 -> 3030: error forwarding port 3030 to pod beaf29b468ce3d61de1e6d046af8b5445495d1443c7c2dfd9cffb9ac03a5b482, uid : exit status 1: 2019/08/27 10:36:51 socat[12993] E connect(5, AF=2 127.0.0.1:3030, 16): Connection refused
[...]
[!]  Helm Operator is not ready yet (Get http://127.0.0.1:58983/healthz: EOF), retrying ...
ERROR: logging before flag.Parse: E0827 19:37:10.087267   16191 portforward.go:331] an error occurred forwarding 58983 -> 3030: error forwarding port 3030 to pod beaf29b468ce3d61de1e6d046af8b5445495d1443c7c2dfd9cffb9ac03a5b482, uid : exit status 1: 2019/08/27 10:37:10 socat[13055] E connect(5, AF=2 127.0.0.1:3030, 16): Connection refused
[!]  Helm Operator is not ready yet (Get http://127.0.0.1:58983/healthz: EOF), retrying ...
[ℹ]  Helm Operator started successfully
[ℹ]  see https://docs.fluxcd.io/projects/helm-operator for details on how to use the Helm Operator
[ℹ]  Waiting for Flux to start
[ℹ]  Flux started successfully
[ℹ]  see https://docs.fluxcd.io/projects/flux for details on how to use Flux
[ℹ]  Committing and pushing manifests to git@github.com:marccarre/my-gitops-repo.git
[master 23e5094] Add Initial Flux configuration
 1 file changed, 38 insertions(+), 38 deletions(-)
 rewrite flux/tiller-ca-cert-configmap.yaml (87%)
Enumerating objects: 7, done.
Counting objects: 100% (7/7), done.
Delta compression using up to 8 threads
Compressing objects: 100% (4/4), done.
Writing objects: 100% (4/4), 2.15 KiB | 2.15 MiB/s, done.
Total 4 (delta 1), reused 0 (delta 0)
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To github.com:marccarre/my-gitops-repo.git
   8535eb4..23e5094  master -> master
[ℹ]  Flux will only operate properly once it has write-access to the Git repository
[ℹ]  please configure git@github.com:marccarre/my-gitops-repo.git so that the following Flux SSH public key has write access to it
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDlp4wLL67rDn6EsNq4LOUbmx7SQ3taOoMl0cykQJf+AFqhEGDLHR9CrtNnoAsAkqiuykrEdj3c1k3a1I0p+o2+hp0iimWa/t9Q+woSYXv3TfNhnw9mMfM6HKNE/7m63j3aUg5o1nsM2fm+Rs67orw4mtjxML9PJJ/Zmq8mDsWNH93Q6WKMwdi2JHhyXVZu0CR0IAzqPU3TQ6McPlZ+coiQtlSJzf9gTu8ImkYK+J4nLvo40BNsFgCNbyxVzHUN9zfjoeMUhhRvj6YRVLBkE06nfxzP6t/UdflHTjfatQR137Pex43HQ6+RV/C6gcX0n3rm8vaIwiSaHb1LimZ/Maiv
```

### CODE REVIEW

File: pkg/gitops/flux/installer.go [hunk 1]
```diff
@@ -133,6 +135,28 @@ func NewInstaller(ctx context.Context, k8sRestConfig *rest.Config, k8sClientSet
...
+		cont, err := continueInstallation()
+		if err != nil {
+			return err
+		}
+		if !cont {
+			logger.Info("Installation of Flux was aborted.")
+			return nil
+		}
+		if cont {
+			err := kubernetes.DeleteNamespace(fi.k8sClientSet, fi.opts.Namespace)
```
Comment (2opremio) [inline pkg/gitops/flux/installer.go]: Shouldn't we also be deleting the roles and helmrelease CRD?

File: pkg/gitops/flux/installer.go [hunk 2]
```diff
@@ -219,6 +243,27 @@ func (fi *Installer) Run(ctx context.Context) error {
 	return nil
 }
 
+func continueInstallation() (bool, error) {
+	for {
+		reader := bufio.NewReader(os.Stdin)
+		fmt.Print("Continue? [y/n]: ")
```
Comment (2opremio) [inline pkg/gitops/flux/installer.go]: Uhm, I am not sure we should be ask the user about this. I would lean towards:

1. Deleting every resource we were to create first (as opposed to deleting the whole namespace since it can contain more stuff)
2. Do it without asking for permission (we shouldn't ask for permission when using `kubectl apply` which would have an overall similar impact).

File: pkg/gitops/flux/installer.go (hunk 1 above)
Comment (marccarre) [reply pkg/gitops/flux/installer.go]: We could, though in practice, there is no conflict for these, given their simplicity, hence I didn't bother. Indeed `CreateOrUpdate` deals with them perfectly well:
```console
[ℹ]  replaced "CustomResourceDefinition.apiextensions.k8s.io/helmreleases.helm.fluxcd.io"
[...]
[ℹ]  replaced "ClusterRoleBinding.rbac.authorization.k8s.io/tiller"
[...]
[ℹ]  replaced "kube-system:RoleBinding.rbac.authorization.k8s.io/tiller-user-binding"
[...]
[ℹ]  replaced "ClusterRole.rbac.authorization.k8s.io/flux"
[ℹ]  replaced "ClusterRoleBinding.rbac.authorization.k8s.io/flux"
[...]
[ℹ]  replaced "ClusterRole.rbac.authorization.k8s.io/flux-helm-operator"
[ℹ]  replaced "ClusterRoleBinding.rbac.authorization.k8s.io/flux-helm-operator"
```

File: pkg/gitops/flux/installer.go (hunk 2 above)
Comment (2opremio) [reply pkg/gitops/flux/installer.go]: Also, I wouldn't overwrite the ssh-key secret (`kubectl apply` would leave it as is, since it's empty in the manifest and it's convenient not needing to overwrite the ssh key every time you reinstall).

Comment (marccarre) [reply pkg/gitops/flux/installer.go]: Grrr, 1. is what I had initially 🤦‍♂

Comment (marccarre) [reply pkg/gitops/flux/installer.go]: Good point about the SSH key 👍

File: pkg/kubernetes/namespace.go [hunk 3]
```diff
@@ -53,6 +55,225 @@ func CheckNamespaceExists(clientSet Interface, name string) (bool, error) {
...
+	// deleting the namespace, hence we wait for commonly used resources.
+	// This isn't perfect in any way, but given there is no way to directly
+	// list all namespaced resources without using kubectl [1] this will
+	// hopefully do in most cases.
+	// [1]: kubectl api-resources --verbs=list --namespaced -o name \
+	//      | xargs -n 1 kubectl get --show-kind --ignore-not-found -n <namespace>
+	return waitForCommonNamespacedResourcesDeletion(clientSet, namespace)
+}
+
+func waitForCommonNamespacedResourcesDeletion(clientSet Interface, namespace string) error {
```
Comment (2opremio) [inline pkg/kubernetes/namespace.go]: FYI, Can't you get this exact effect using the `"Foreground"` DeletionPropagation policy when invoking delete on the namespace? https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#DeletionPropagation

File: pkg/gitops/flux/installer.go [hunk 4]
```diff
@@ -133,6 +135,28 @@ func NewInstaller(ctx context.Context, k8sRestConfig *rest.Config, k8sClientSet
 
 // Run runs the Flux installer
 func (fi *Installer) Run(ctx context.Context) error {
+	fluxNSExists, err := kubernetes.CheckNamespaceExists(fi.k8sClientSet, fi.opts.Namespace)
+	if err != nil {
+		return errors.Wrapf(err, "cannot check if namespace %s exists", fi.opts.Namespace)
+	}
+	if fluxNSExists {
+		logger.Info("Namespace \"%s\" already exists. eksctl will now delete all its content before proceeding to install Flux.", fi.opts.Namespace)
```
Comment (errordeveloper) [inline pkg/gitops/flux/installer.go]: Our current style convention for log messages dictates that they always start with a lower-case character (unless a name is used).

File: pkg/gitops/flux/installer.go (hunk 1 above)
Comment (errordeveloper) [reply pkg/gitops/flux/installer.go]: This bring the topic of "what constitutes a component?". Proper way to do it would be via [Application CRD](https://github.com/kubernetes-sigs/application), I strongly recommend that we adopt it going forward.

File: pkg/gitops/flux/installer.go (hunk 2 above)
Comment (errordeveloper) [reply pkg/gitops/flux/installer.go]: As I mentioned in slack, we should always assume non-interactive use-cases. None of other commands prompt user for anything. This is one of the things I'd like to add to CLI design guidelines. We should make softer behaviour  the default and provide a flag to override (but only if there is strong use-case for such flag).

**Comment by errordeveloper** [comment]
I wonder if we should just say that we own the flux namespace, and just delete the whole namespace.

File: pkg/kubernetes/namespace.go (hunk 3 above)
Comment (marccarre) [reply pkg/kubernetes/namespace.go]: Maybe, will look into it. Thanks for the pointer 👍

File: pkg/gitops/flux/installer.go (hunk 4 above)
Comment (marccarre) [reply pkg/gitops/flux/installer.go]: Noted, will fix. Thanks 👍

**Comment by marccarre** [comment]
> I wonder if we should just say that we own the flux namespace, and just delete the whole namespace.

@errordeveloper this PR _does_ delete the entire namespace, however, there is a good (at least in my opinion) argument in favour of not deleting everything here: https://github.com/weaveworks/eksctl/pull/1203#discussion_r318019005

**Comment by errordeveloper** [comment]
> ...there is a good (at least in my opinion) argument in favour of not deleting everything here: #1203 (comment)

@marccarre I'd like to discuss the SSH key separately. My view is that it we shouldn't rely on Flux to generate the key.

The reason to that is the following:

We will definitely need a way for user to specify the key file or body. We can do that in way where we have two different modes - key is either supplied and passed in, or it's generated by flux.
If there was no extra complexity, I wouldn't mind that, however obtaining the key from flux requires forwarding a port, which we currently do, but I would rather get rid of this complexity and allow users to disable flux daemon port (possibly make that the default).
So I think the design would be much less complex if we have two modes:
- user provides the key, eksctl creates the secret
- eksctl generates the key, writes it to a file locally and creates the secret

Aside from that, to your point, we can delete namespace as they key is defined by eksctl.

**Comment by errordeveloper** [comment]
I am not suggesting to change the behaviour of key generation right now, we should discuss that topic separately.

**Comment by marccarre** [comment]
I've opened https://github.com/weaveworks/eksctl/issues/1213 to discuss the question of the SSH key further. For now, I'll close this PR, and will soon open another one specifically deleting the Flux resources (except the SSH key) and nothing else, in order to avoid accidentally deleting anything else belonging to the end-user, as suggested by @2opremio and voted [here](https://weaveworks.slack.com/archives/CJE5R99FV/p1566903889036200).

File: pkg/kubernetes/namespace.go (hunk 3 above)
Comment (marccarre) [reply pkg/kubernetes/namespace.go]: It may work when doing so on the namespace (still haven't tried), but for the record, it doesn't help now that we delete individual resources. 😞
